extensions/markdown/media/main.js

```diff
@@ -5,16 +5,114 @@
 
 'use strict';
 
-let pageHeight = 0;
-
-window.onload = () => {
-	pageHeight = document.body.getBoundingClientRect().height;
-};
-
-window.addEventListener('resize', () => {
-	const currentOffset = window.scrollY;
-	const newPageHeight = document.body.getBoundingClientRect().height;
-	const dHeight = newPageHeight / pageHeight;
-	window.scrollTo(0, currentOffset * dHeight);
-	pageHeight = newPageHeight;
-}, true);
+(function () {
+	var pageHeight = 0;
+
+	/**
+	 * Find the elements around line.
+	 *
+	 * If an exact match, returns a single element. If the line is between elements,
+	 * returns the element before and the element after the given line.
+	 */
+	function getElementsAroundSourceLine(targetLine) {
+		const lines = document.getElementsByClassName('code-line');
+		let before = null;
+		for (const element of lines) {
+			const lineNumber = +element.getAttribute('data-line');
+			if (isNaN(lineNumber)) {
+				continue;
+			}
+			const entry = { line: lineNumber, element: element };
+			if (lineNumber === targetLine) {
+				return { before: entry, after: null };
+			} else if (lineNumber > targetLine) {
+				return { before, after: entry };
+			}
+			before = entry;
+		}
+		return { before };
+	}
+
+	function getSourceRevealAddedOffset() {
+		return -(window.innerHeight * 1 / 5);
+	}
+
+	/**
+	 * Attempt to reveal the element for a source line in the editor.
+	 */
+	function scrollToRevealSourceLine(line) {
+		const {before, after} = getElementsAroundSourceLine(line);
+		if (before) {
+			let scrollTo = 0;
+			if (after) {
+				// Between two elements. Go to percentage offset between them.
+				const betweenProgress = (line - before.line) / (after.line - before.line);
+				const elementOffset = after.element.getBoundingClientRect().top - before.element.getBoundingClientRect().top;
+				scrollTo = before.element.getBoundingClientRect().top + betweenProgress * elementOffset;
+			} else {
+				scrollTo = before.element.getBoundingClientRect().top;
+			}
+			window.scroll(0, window.scrollY + scrollTo + getSourceRevealAddedOffset());
+		}
+	}
+
+	function didUpdateScrollPosition(offset) {
+		const lines = document.getElementsByClassName('code-line');
+		let nearest = lines[0];
+		for (let i = lines.length - 1; i >= 0; --i) {
+			const lineElement = lines[i];
+			if (offset <= window.scrollY + lineElement.getBoundingClientRect().top + lineElement.getBoundingClientRect().height) {
+				nearest = lineElement;
+			} else {
+				break;
+			}
+		}
+
+		if (nearest) {
+			const line = +nearest.getAttribute('data-line');
+			const args = [window.initialData.source, line];
+			window.parent.postMessage({
+				command: "did-click-link",
```

Comment:
This functionality is something we would probably want to standardize, similar to the `postMessage` command. Emulating clicks is pretty iffy.